rustler_codegen/src/tagged_enum.rs

```diff
@@ -0,0 +1,337 @@
+                        Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
+                                        "Could not decode field :{:?} on %{{}}",
+                                        field
+                        )))),
+                        Ok(value) => Ok(value),
+                    }
+                }
+
+                #(#variant_defs)*
+
+                Err(::rustler::Error::RaiseAtom("invalid_variant"))
+            }
+        }
+    };
+
+    gen
+}
+
+fn gen_encoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) -> TokenStream {
+    let enum_type = &ctx.ident_with_lifetime;
+    let enum_name = ctx.ident;
+
+    let variant_defs: Vec<TokenStream> = variants
+        .iter()
+        .map(|variant| {
+            let variant_ident = &variant.ident;
+            let atom_fn = Context::ident_to_atom_fun(variant_ident);
+
+            match &variant.fields {
+                Fields::Unit => gen_unit_encoder(enum_name, variant_ident, atom_fn),
+                Fields::Unnamed(fields) => {
+                    gen_unnamed_encoder(enum_name, fields, variant_ident, atom_fn)
+                }
+                Fields::Named(fields) => {
+                    gen_named_encoder(enum_name, fields, variant_ident, atom_fn)
+                }
+            }
+        })
+        .collect();
+
+    let gen = quote! {
+        impl<'b> ::rustler::Encoder for #enum_type {
+            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
+                use #atoms_module_name::*;
+
+                match self {
+                    #(#variant_defs)*
+                }
+            }
+        }
+    };
+
+    gen
+}
+
+fn gen_unit_decoder(enum_name: &Ident, variant_ident: &Ident, atom_fn: Ident) -> TokenStream {
+    quote! {
+        if let Ok(true) = value.as_ref().map(|a| *a == #atom_fn()) {
+            return Ok ( #enum_name :: #variant_ident );
+        }
+    }
+}
+
+fn gen_unnamed_decoder<'a>(
+    enum_name: &Ident,
+    fields: &FieldsUnnamed,
+    fields_iter: impl Iterator<Item = &'a Field>,
+    variant_ident: &Ident,
+    atom_fn: Ident,
+) -> TokenStream {
+    let decoded_field = &fields_iter
+        .enumerate()
+        .map(|(i, f)| {
+            let i = i + 1;
+            let ty = &f.ty;
+            quote! {
+                <#ty>::decode(tuple[#i]).map_err(|_| ::rustler::Error::RaiseTerm(
+                    Box::new(format!("Could not decode field on position {}", #i))
+                ))?
+            }
+        })
+        .collect::<Vec<_>>();
+    let len = fields.unnamed.len();
+    quote! {
+        if let Ok(tuple) = ::rustler::types::tuple::get_tuple(term) {
+            let name = tuple
+                .get(0)
+                .and_then(|&first| ::rustler::types::atom::Atom::from_term(first).ok())
+                .ok_or(::rustler::Error::RaiseAtom("invalid_variant"))?;
+            if name == #atom_fn() {
+                if tuple.len() - 1 != #len {
```

Comment:
I think we don't need that check, do we? If we cannot decode the field on index `i`, the decoder further above will report an error.

Reply:
(Or does this somehow behave differently then the named case?)

Reply:
Scratch that, the error message probably already indicates to the user quite well what the problem is!